### PR TITLE
Remaster the WakeConfig API

### PR DESCRIPTION
**Pull request by JakeSiFive**
This change significantly refactors the WakeConfig API with the following benefits

1) New fields can be easily added by simply adding a policy with much less boiler plate or surgery
2) Fields not have automatically tracked provenance (crap I need to go re-spell that everywhere now) so we can specify where the config field came from when printing out the config
3) There is now support for command line overrides though nothing currently uses it

### CODE REVIEW

File: src/runtime/config.h [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    case WakeConfigProvenance::Default:
+      return "Default";
+    case WakeConfigProvenance::Wakeroot:
+      return "Wakeroot";
+    case WakeConfigProvenance::Userconfig:
+      return "Userconfig";
+    case WakeConfigProvenance::CommandLine:
+      return "Commandline";
+  }
+  return "Default";
```
Comment (V-FEXrt) [inline src/runtime/config.h:77]: nit: `return "Unknown";`

File: src/runtime/config.h [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+  static void_t add_userconfig_key(std::set<std::string>& out) {
+    if (P::allowed_in_userconfig) {
+      out.emplace(P::key);
+    }
+    return {};
+  }
+
+  template <WakeConfigProvenance p, class P>
+  void_t set_policy(const JAST& json) {
+    // The current semantics want us to
```
Comment (V-FEXrt) [inline src/runtime/config.h:123]: This comment is unclear.  Needs to be more detailed or removed

**Review by V-FEXrt — CHANGES_REQUESTED** [review]

File: src/runtime/config.cpp [hunk 3]
```diff
@@ -277,44 +296,32 @@ bool init(const std::string& wakeroot_path) {
...
-  // {
-  //   auto json_version = user_config_json.expect_string("version");
-  //   if (json_version) {
-  //     version = *json_version;
-  //   }
-  // }
-
-  static WakeConfig wc = {version, user_config_path};
-  _config = &wc;
+  _config->set_all<WakeConfigProvenance::Userconfig>(user_config_json);
```
[the remark this answers is not in the document]
Comment (JakeSiFive) [reply src/runtime/config.cpp:308]: Why?

File: src/runtime/config.h [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+  static constexpr bool allowed_in_userconfig = false;
+  std::string version = "";
+  static constexpr std::string VersionPolicy::*value = &VersionPolicy::version;
+  static constexpr Override<std::string> override_value = nullptr;
+
+  VersionPolicy() = default;
+  static void set(VersionPolicy& p, const JAST& json);
+};
+
+struct UserConfigPolicy {
```
[the remark this answers is not in the document]
Comment (JakeSiFive) [reply src/runtime/config.h:48]: There's functionality baked into this change for specific overrides but I suppose I could also add a generic override as well in a follow up change. In fact if we had generic overrides do we even want specific overrides? Maybe I should change this to work exclusively with generic overrides now?

File: src/runtime/config.h [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+  return "Default";
+}
+
+template <class... Polcies>
+struct WakeConfigImpl : public Polcies... {
+  WakeConfigImpl(const WakeConfigImpl&) = delete;
+  WakeConfigImpl(WakeConfigImpl&&) = delete;
+  WakeConfigImpl() = default;
+
+  std::map<std::string, WakeConfigProvenance> provinence;
```
[the remark this answers is not in the document]
Comment (JakeSiFive) [reply src/runtime/config.h:86]: The one without the 'i' sorry didn't catch all of them

File: src/runtime/config.h (hunk 2 above)
Comment (JakeSiFive) [reply src/runtime/config.h:123]: Yeah it was half written lol

File: src/runtime/config.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (JakeSiFive) [reply src/runtime/config.cpp:308]: Yeah I'm quite certain this is just wrong

[the remark this answers is not in the document]
Comment (V-FEXrt) [reply src/runtime/config.cpp:308]: My complaint it with `Userconfig` vs `UserConfig`

[the remark this answers is not in the document]
Comment (V-FEXrt) [reply src/runtime/config.cpp:308]: Two words in an enum should be CamelCase

[the remark this answers is not in the document]
Comment (JakeSiFive) [reply src/runtime/config.cpp:308]: oh gotcha. The much bigger change there was the `WakeConfigProvenance::` part lol

**Review by V-FEXrt — APPROVED** [review]

**Pull request merged**